kartothek/io_components/write.py

```diff
@@ -246,3 +253,28 @@ def raise_if_dataset_exists(dataset_uuid, store):
                 )
     except KeyError:
         pass
+
+
+def check_dataset_creation_constraints(
+    factory, partition_on, secondary_indices, overwrite=False
+):
+    """Check if all required constraints for creating a dataset are met."""
+
+    if overwrite is not False:
+        warnings.warn(
+            "The keyword `overwrite` be removed in the next major release ",  # FIXME: keep or not?
+            DeprecationWarning,
+            stacklevel=2,
+        )
+        if DatasetMetadata.exists(
+            uuid=factory.dataset_uuid, store=factory.store_factory()
+        ):
+            # delete common metadata after partitions
+            delete_common_metadata(dataset_factory=factory)
+            delete_indices(dataset_factory=factory)
+            # Delete the top level metadata file
+            delete_top_level_metadata(dataset_factory=factory)
```

Comment:
This changes functionality quite significantly. I'm quite strongly against deleting anything explicitly in a write operation. If the write operation doesn't success for whatever reason (e.g. final commit doesn't succeed due to schema violations) the previous implementation kept the dataset intact.

At the moment there is the potential for an inconsistent dataset state in case the dataset common metadata has been written but the rest of the commit failed. I'm willing to accept this fringe edge case in favour of protection from data loss.

Reply:
> If the write operation doesn't success for whatever reason (e.g. final commit doesn't succeed due to schema violations) the previous implementation kept the dataset intact.

Hmm, right. 

- [ ] look into this